README.md
shadowsocks manyuser branch
===========
Which people need this branch
------------------
1.share shadowsocks server

2.create multi server by shadowsocks

3.manage server (transfer / account)

Change
-------
从efd106dfb5这个提交开始能够独立于前端运行了，原因在于考虑到udp的话可靠性不好。在数据库中修改passwd，switch，enable，流量都会立即生效(比如改密码不需要再通过Managesocket去手动stop某个服务了)。不过内部实现还是通过数据库线程发送一个udp包来触发的，因为数据库这个操作改成异步的会灰常灰常麻烦，所以千万不要在iptabls里面把udp全给堵死了。<del>我的线上环境变成了瘟都死所以这个没有时间和环境测试，对运行过程中出现服务崩溃，服务器冒烟不负任何责任。</del>发现有问题的话给我个提交？诶哟好累英语太拙计就写中文好了。

Install
-------
install MySQL 5.x.x

`pip install cymysql`

create a database named `shadowsocks`

import `shadowsocks.sql` into `shadowsocks`

edit Config.py
Example:

	#Config
	MYSQL_HOST = 'mdss.mengsky.net'
	MYSQL_PORT = 3306
	MYSQL_USER = 'ss'
	MYSQL_PASS = 'ss'
	MYSQL_DB = 'shadowsocks'

	MANAGE_PASS = 'ss233333333'
	#if you want manage in other server you should set this value to global ip
	MANAGE_BIND_IP = '127.0.0.1'
	#make sure this port is idle
	MANAGE_PORT = 23333

TestRun `cd shadowsocks` ` python server.py`

if no exception server will startup. you will see such like
Example:

	db start server at port [%s] pass [%s]

Database user table column
------------------
`passwd` server pass

`port` server port

`t` last keepalive time

`u` upload transfer

`d` download transer

`transfer_enable` if u + d > transfer_enable this server will be stop (db_transfer.py del_server_out_of_bound_safe)

Manage socket
------------------
Manage server work in UDP at `MANAGE_BIND_IP` `MANAGE_PORT`

use `MANAGE_PASS:port:passwd:0` to del a server at port `port`

use `MANAGE_PASS:port:passwd:1` to run a server at port `port` password is `passwd`

Python Eg:

	udpsock.sendto('MANAGE_PASS:65535:123456:1', (MANAGE_BIND_IP, MANAGE_PORT))
	
PHP Eg:

	$sock = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
	$msg = 'MANAGE_PASS:65535:123456:1';
	$len = strlen($msg);
	socket_sendto($sock, $msg, $len, 0, MANAGE_BIND_IP, MANAGE_PORT);
	socket_close($sock);

NOTICE
------------------
If error such like `2014-09-18 09:02:37 ERROR    [Errno 24] Too many open files`

edit /etc/security/limits.conf

Add:

	*                soft    nofile          8192
	*                hard    nofile          65535


add `ulimit -n 8192` in your startup script before runing

shadowsocks
===========

[![PyPI version]][PyPI] [![Build Status]][Travis CI] 

A fast tunnel proxy that helps you bypass firewalls.

[中文说明][Chinese Readme]

Install
-------

You'll have a client on your local machine, and install a server on a
remote server.

### Client

* [Windows] / [OS X]
* [Android] / [iOS]
* [OpenWRT]

### Server

#### Debian / Ubuntu:

    apt-get install python-pip python-m2crypto
    pip install shadowsocks

Or simply `apt-get install shadowsocks` if you have [Debian sid] in your
source list.

#### CentOS:

    yum install m2crypto python-setuptools
    easy_install pip
    pip install shadowsocks

Configuration
-------------

On your server create a config file `/etc/shadowsocks.json`.
Example:

    {
        "server":"my_server_ip",
        "server_port":8388,
        "local_address": "127.0.0.1",
        "local_port":1080,
        "password":"mypassword",
        "timeout":300,
        "method":"aes-256-cfb",
        "fast_open": false,
        "workers": 1
    }

Explanation of the fields:

| Name          | Explanation                                     |
| ------------- | ----------------------------------------------- |
| server        | the address your server listens                 |
| server_port   | server port                                     |
| local_address | the address your local listens                  |
| local_port    | local port                                      |
| password      | password used for encryption                    |
| timeout       | in seconds                                      |
| method        | default: "aes-256-cfb", see [Encryption]        |
| fast_open     | use [TCP_FASTOPEN], true / false                |
| workers       | number of workers, available on Unix/Linux      |

Run `ssserver -c /etc/shadowsocks.json` on your server. To run it in the
background, use [Supervisor].

On your client machine, use the same configuration as your server, and
start your client.

If you use Chrome, it's recommended to use [SwitchySharp]. Change the proxy 
settings to

    protocol: socks5
    hostname: 127.0.0.1
    port:     your local_port

If you can't install [SwitchySharp], you can launch Chrome with the following
arguments to force Chrome to use the proxy:

    Chrome.exe --proxy-server="socks5://127.0.0.1:1080" --host-resolver-rules="MAP * 0.0.0.0 , EXCLUDE localhost"

If you can't even download Chrome, find a friend to download a
[Chrome Standalone] installer for you.

Command line args
------------------

You can use args to override settings from `config.json`.

    sslocal -s server_name -p server_port -l local_port -k password -m bf-cfb
    ssserver -p server_port -k password -m bf-cfb --workers 2
    ssserver -c /etc/shadowsocks/config.json

List all available args with `-h`.

Wiki
----

https://github.com/clowwindy/shadowsocks/wiki

License
-------
MIT

Bugs and Issues
----------------

* [Troubleshooting]
* [Issue Tracker]
* [Mailing list]


[Android]:           https://github.com/clowwindy/shadowsocks/wiki/Ports-and-Clients#android
[Build Status]:      https://img.shields.io/travis/clowwindy/shadowsocks/master.svg?style=flat
[Chinese Readme]:    https://github.com/clowwindy/shadowsocks/wiki/Shadowsocks-%E4%BD%BF%E7%94%A8%E8%AF%B4%E6%98%8E
[Chrome Standalone]: https://support.google.com/installer/answer/126299
[Debian sid]:        https://packages.debian.org/unstable/python/shadowsocks
[Encryption]:        https://github.com/clowwindy/shadowsocks/wiki/Encryption
[iOS]:               https://github.com/shadowsocks/shadowsocks-iOS/wiki/Help
[Issue Tracker]:     https://github.com/clowwindy/shadowsocks/issues?state=open
[Mailing list]:      http://groups.google.com/group/shadowsocks
[OpenWRT]:           https://github.com/clowwindy/shadowsocks/wiki/Ports-and-Clients#openwrt
[OS X]:              https://github.com/shadowsocks/shadowsocks-iOS/wiki/Shadowsocks-for-OSX-Help
[PyPI]:              https://pypi.python.org/pypi/shadowsocks
[PyPI version]:      https://img.shields.io/pypi/v/shadowsocks.svg?style=flat
[Supervisor]:        https://github.com/clowwindy/shadowsocks/wiki/Configure-Shadowsocks-with-Supervisor
[TCP_FASTOPEN]:      https://github.com/clowwindy/shadowsocks/wiki/TCP-Fast-Open
[Travis CI]:         https://travis-ci.org/clowwindy/shadowsocks
[Troubleshooting]:   https://github.com/clowwindy/shadowsocks/wiki/Troubleshooting
[SwitchySharp]:      https://chrome.google.com/webstore/detail/proxy-switchysharp/dpplabbmogkhghncfbfdeeokoefdjegm
[Windows]:           https://github.com/clowwindy/shadowsocks/wiki/Ports-and-Clients#windows


CHANGES
2.2.0 2014-09-09
- Add RC4-MD5 encryption

2.1.0 2014-08-10
- Use only IPv4 DNS server
- Does not ship config.json
- Better error message

2.0.12 2014-07-26
- Support -q quiet mode
- Exit 0 when showing help with -h

2.0.11 2014-07-12
- Prefers IP addresses over hostnames, more friendly with socksify and openvpn

2.0.10 2014-07-11
- Fix UDP on local

2.0.9 2014-07-06
- Fix EWOULDBLOCK on Windows
- Fix Unicode config problem on some platforms

2.0.8 2014-06-23
- Use multiple DNS to query hostnames

2.0.7 2014-06-21
- Fix fastopen on local
- Fallback when fastopen is not available
- Add verbose logging mode -vv
- Verify if hostname is valid

2.0.6 2014-06-19
- Fix CPU 100% on POLL_HUP
- More friendly logging

2.0.5 2014-06-18
- Support a simple config format for multiple ports

2.0.4 2014-06-12
- Fix worker master

2.0.3 2014-06-11
- Fix table encryption with UDP

2.0.2 2014-06-11
- Add asynchronous DNS in TCP relay

2.0.1 2014-06-05
- Better logging
- Maybe fix bad file descriptor

2.0 2014-06-05
- Use a new event model
- Remove gevent
- Refuse to use default password
- Fix a problem when using multiple passwords with table encryption

1.4.5 2014-05-24
- Add timeout in TCP server
- Close sockets in master process

1.4.4 2014-05-17
- Support multiple workers

1.4.3 2014-05-13
- Fix Windows

1.4.2 2014-05-10
- Add salsa20-ctr cipher

1.4.1 2014-05-03
- Fix error log
- Fix EINPROGESS with some version of gevent

1.4.0 2014-05-02
- Adds UDP relay
- TCP fast open support on Linux 3.7+

1.3.7 2014-04-10
- Fix a typo in help

1.3.6 2014-04-10
- Fix a typo in help

1.3.5 2014-04-07
- Add help
- Change default local binding address into 127.0.0.1

1.3.4 2014-02-17
- Fix a bug when no config file exists
- Client now support multiple server ports and multiple server/port pairs
- Better error message with bad config.json format and wrong password

1.3.3 2013-07-09
- Fix default key length of rc2

1.3.2 2013-07-04
- Server will listen at server IP specified in config
- Check config file and show some warning messages

1.3.1 2013-06-29
- Fix -c arg

1.3.0 2013-06-22
- Move to pypi

1.2.3 2013-06-14
- add bind address

1.2.2 2013-05-31
- local can listen at ::0 with -6 arg; bump 1.2.2

1.2.1 2013-05-23
- Fix an OpenSSL crash

1.2 2013-05-22
- Use random iv, we finally have strong encryption

1.1.1 2013-05-21
- Add encryption, AES, blowfish, etc.

1.1 2013-05-16
- Support IPv6 addresses (type 4)
- Drop Python 2.5 support

1.0 2013-04-03
- Fix -6 IPv6

0.9.4 2013-03-04
- Support Python 2.5

0.9.3 2013-01-14
- Fix conn termination null data

0.9.2 2013-01-05
- Change default timeout

0.9.1 2013-01-05
- Add Travis-CI test

0.9 2012-12-30
- Replace send with sendall, fix FreeBSD

0.6 2012-12-06
- Support args

0.5 2012-11-08
- Fix encryption with negative md5sum

0.4 2012-11-02
- Move config into a JSON file
- Auto-detect config path

0.3 2012-06-06
- Move socks5 negotiation to local

0.2 2012-05-11
- Add -6 arg for IPv6
- Fix socket.error

0.1 2012-04-20
- Initial version


shadowsocks/utils.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright (c) 2014 clowwindy
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import os
import json
import sys
import getopt
import logging


VERBOSE_LEVEL = 5


def check_python():
    info = sys.version_info
    if not (info[0] == 2 and info[1] >= 6):
        print 'Python 2.6 or 2.7 required'
        sys.exit(1)


def print_shadowsocks():
    version = ''
    try:
        import pkg_resources
        version = pkg_resources.get_distribution('shadowsocks').version
    except Exception:
        pass
    print 'shadowsocks %s' % version


def find_config():
    config_path = 'config.json'
    if os.path.exists(config_path):
        return config_path
    config_path = os.path.join(os.path.dirname(__file__), '../', 'config.json')
    if os.path.exists(config_path):
        return config_path
    return None


def check_config(config):
    if config.get('local_address', '') in ['0.0.0.0']:
        logging.warn('warning: local set to listen 0.0.0.0, which is not safe')
    if config.get('server', '') in ['127.0.0.1', 'localhost']:
        logging.warn('warning: server set to listen %s:%s, are you sure?' %
                     (config['server'], config['server_port']))
    if (config.get('method', '') or '').lower() == '':
        logging.warn('warning: table is not safe; please use a safer cipher, '
                     'like AES-256-CFB')
    if (config.get('method', '') or '').lower() == 'rc4':
        logging.warn('warning: RC4 is not safe; please use a safer cipher, '
                     'like AES-256-CFB')
    if config.get('timeout', 300) < 100:
        logging.warn('warning: your timeout %d seems too short' %
                     int(config.get('timeout')))
    if config.get('timeout', 300) > 600:
        logging.warn('warning: your timeout %d seems too long' %
                     int(config.get('timeout')))
    if config.get('password') in ['mypassword', 'barfoo!']:
        logging.error('DON\'T USE DEFAULT PASSWORD! Please change it in your '
                      'config.json!')
        exit(1)


def get_config(is_local):
    logging.basicConfig(level=logging.INFO,
                        format='%(levelname)-s: %(message)s', filemode='a+')
    if is_local:
        shortopts = 'hs:b:p:k:l:m:c:t:vq'
        longopts = ['fast-open']
    else:
        shortopts = 'hs:p:k:m:c:t:vq'
        longopts = ['fast-open', 'workers:']
    try:
        config_path = find_config()
        optlist, args = getopt.getopt(sys.argv[1:], shortopts, longopts)
        for key, value in optlist:
            if key == '-c':
                config_path = value

        if config_path:
            logging.info('loading config from %s' % config_path)
            with open(config_path, 'rb') as f:
                try:
                    config = json.load(f, object_hook=_decode_dict)
                except ValueError as e:
                    logging.error('found an error in config.json: %s',
                                  e.message)
                    sys.exit(1)
        else:
            config = {}

        optlist, args = getopt.getopt(sys.argv[1:], shortopts, longopts)
        v_count = 0
        for key, value in optlist:
            if key == '-p':
                config['server_port'] = int(value)
            elif key == '-k':
                config['password'] = value
            elif key == '-l':
                config['local_port'] = int(value)
            elif key == '-s':
                config['server'] = value
            elif key == '-m':
                config['method'] = value
            elif key == '-b':
                config['local_address'] = value
            elif key == '-v':
                v_count += 1
                # '-vv' turns on more verbose mode
                config['verbose'] = v_count
            elif key == '-t':
                config['timeout'] = int(value)
            elif key == '--fast-open':
                config['fast_open'] = True
            elif key == '--workers':
                config['workers'] = value
            elif key == '-h':
                if is_local:
                    print_local_help()
                else:
                    print_server_help()
                sys.exit(0)
            elif key == '-q':
                v_count -= 1
                config['verbose'] = v_count
    except getopt.GetoptError as e:
        print >>sys.stderr, e
        print_help(is_local)
        sys.exit(2)

    if not config:
        logging.error('config not specified')
        print_help(is_local)
        sys.exit(2)

    config['password'] = config.get('password', None)
    config['method'] = config.get('method', 'aes-256-cfb')
    config['port_password'] = config.get('port_password', None)
    config['timeout'] = int(config.get('timeout', 300))
    config['fast_open'] = config.get('fast_open', False)
    config['workers'] = config.get('workers', 1)
    config['verbose'] = config.get('verbose', False)
    config['local_address'] = config.get('local_address', '127.0.0.1')
    config['local_port'] = config.get('local_port', 1080)
    if is_local:
        if config.get('server', None) is None:
            logging.error('server addr not specified')
            print_local_help()
            sys.exit(2)
    else:
        config['server'] = config.get('server', '0.0.0.0')
    config['server_port'] = config.get('server_port', 8388)

    if not ('password' in config and config['password']):
        logging.error('password not specified')
        print_help(is_local)
        sys.exit(2)

    logging.getLogger('').handlers = []
    logging.addLevelName(VERBOSE_LEVEL, 'VERBOSE')
    if config['verbose'] >= 2:
        level = VERBOSE_LEVEL
    elif config['verbose'] == 1:
        level = logging.DEBUG
    elif config['verbose'] == -1:
        level = logging.WARN
    elif config['verbose'] <= -2:
        level = logging.ERROR
    else:
        level = logging.INFO
    logging.basicConfig(level=level,
                        format='%(asctime)s %(levelname)-8s %(message)s',
                        datefmt='%Y-%m-%d %H:%M:%S', filemode='a+')

    check_config(config)

    return config


def print_help(is_local):
    if is_local:
        print_local_help()
    else:
        print_server_help()


def print_local_help():
    print '''usage: sslocal [-h] -s SERVER_ADDR [-p SERVER_PORT]
               [-b LOCAL_ADDR] [-l LOCAL_PORT] -k PASSWORD [-m METHOD]
               [-t TIMEOUT] [-c CONFIG] [--fast-open] [-v] [-q]

optional arguments:
  -h, --help            show this help message and exit
  -s SERVER_ADDR        server address
  -p SERVER_PORT        server port, default: 8388
  -b LOCAL_ADDR         local binding address, default: 127.0.0.1
  -l LOCAL_PORT         local port, default: 1080
  -k PASSWORD           password
  -m METHOD             encryption method, default: aes-256-cfb
  -t TIMEOUT            timeout in seconds, default: 300
  -c CONFIG             path to config file
  --fast-open           use TCP_FASTOPEN, requires Linux 3.7+
  -v, -vv               verbose mode
  -q, -qq               quiet mode, only show warnings/errors

Online help: <https://github.com/clowwindy/shadowsocks>
'''


def print_server_help():
    print '''usage: ssserver [-h] [-s SERVER_ADDR] [-p SERVER_PORT] -k PASSWORD
                -m METHOD [-t TIMEOUT] [-c CONFIG] [--fast-open]
                [--workers WORKERS] [-v] [-q]

optional arguments:
  -h, --help            show this help message and exit
  -s SERVER_ADDR        server address, default: 0.0.0.0
  -p SERVER_PORT        server port, default: 8388
  -k PASSWORD           password
  -m METHOD             encryption method, default: aes-256-cfb
  -t TIMEOUT            timeout in seconds, default: 300
  -c CONFIG             path to config file
  --fast-open           use TCP_FASTOPEN, requires Linux 3.7+
  --workers WORKERS     number of workers, available on Unix/Linux
  -v, -vv               verbose mode
  -q, -qq               quiet mode, only show warnings/errors

Online help: <https://github.com/clowwindy/shadowsocks>
'''


def _decode_list(data):
    rv = []
    for item in data:
        if isinstance(item, unicode):
            item = item.encode('utf-8')
        elif isinstance(item, list):
            item = _decode_list(item)
        elif isinstance(item, dict):
            item = _decode_dict(item)
        rv.append(item)
    return rv


def _decode_dict(data):
    rv = {}
    for key, value in data.iteritems():
        if isinstance(key, unicode):
            key = key.encode('utf-8')
        if isinstance(value, unicode):
            value = value.encode('utf-8')
        elif isinstance(value, list):
            value = _decode_list(value)
        elif isinstance(value, dict):
            value = _decode_dict(value)
        rv[key] = value
    return rv
